Magic Rework Part 2

Wollywoger opened the pull request:
This is the second half of the magic rework i finished a **while** ago. This is mostly meant to bring that system further to being integrated with the base game, and adds a number of changes and features to improve this system

List of additions/changes:
- Spellbook is now split between cantrips and manipulation, cantrips are most of all spells, while manipulation is utilities surronding magic
* While some QoL is definitely gonna be needed down the line to make navigating this better, that would fall beyond the scope, and my own skills/knowledge of UX
- Leylines are still in a screwy state, but now i've set it up so it dumps into volite amulets
- Speaking of, the normal/intended transference methods baked in have been fixed

Added two (and a half) sources of mana: 
- The Stump Which Watches, a stump that can be fed flowers from botany to generate mana. Crafted using wood and a volite crystal
- the Stellar Oculory, a machine that when placed in starlight, passively generates mana. Craftable like a normal machine
-  Theres also the meditation mat, an object that empowers all "meditate" type mana charges with a fixed (and higher) mana recharge count
- wands, which are presently used to transfer from magic altars, craftable using pens, rods, cables for a makeshift one: wood and gold or through printing via research for a proper one

 New spells:
- Lesser Splatter casting, an empowered and faster recharging version of mana charge which drains some of your blood when you use it
- Shatter Lights, basically a projectile spell that shoots the fisher bullet that disrupts lights
- Eyes of Night, a temporary night vision spell

- Adds a psionic quirk, which gives you a mana refund on specific spells, currently only mage hand and sending. Also gives you a summonable pseudo-wand.

- fixes and rebalances stuff left in until now (meditation spam bug, volite amulet has been reworked, etc)

I also did some house keeping involving magic stuff, moving them to better folders, and made some mirrored folders for the vars defines and what have you 
notable is that i made the starlight checks that i copied from starlight condensation now available on all subtypes of `/atom/movable` since i anticipate this sort of thing being used by other people

Wollywoger committed:
merges spells into one tab

Wollywoger committed:
first commit, changes meditation

Wollywoger committed:
first main commit

Wollywoger committed:
Merge branch 'MrMelbert:master' into magic-rework-part-2

Wollywoger committed:
progress

Wollywoger committed:
just comitting this to checkout, wands

Wollywoger committed:
big changes, first version of wands

Wollywoger committed:
dme

Wollywoger committed:
temporarily disables a feature so it can compiles

Wollywoger committed:
fixes compilation errors

Wollywoger committed:
wand work good

Wollywoger committed:
adds the newer spells to the cantrip tab

Wollywoger committed:
work trying to fix the null index error

Wollywoger committed:
cleans up stuff from the old tabs, fixes bad index issue

Wollywoger committed:
fixes transferrence. Lol.

Wollywoger committed:
mild changes, first parts of the nature shrine, adds eyes of night to the spellbook list and changes its file name

Wollywoger committed:
fixes eyes of night, makes nature shrine actually work

Wollywoger committed:
first bits of work on oculory, refactors nature shrine

Wollywoger committed:
manually kills has_mana in the dme

Wollywoger committed:
Merge branch 'master' of https://www.github.com/wollywoger/maplestationcode into magic-rework-part-2

Wollywoger committed:
stellar occulory just works now i guess

Wollywoger committed:
removes some debug/redundant code

Wollywoger committed:
Merge branch 'master' into magic-rework-part-2

Wollywoger committed:
adds in code for alt wands + recipes, updates sprite for occulory

Wollywoger committed:
Merge branch 'magic-rework-part-2' of https://www.github.com/wollywoger/maplestationcode into magic-rework-part-2

Wollywoger committed:
Merge branch 'master' into magic-rework-part-2

Wollywoger committed:
repathing, removes the old aeromancy file, touches up wands

Wollywoger committed:
house keeping surronding magic charge spells, fixes up and adds splatter casting

Wollywoger committed:
artificial volite, compile fixes

Wollywoger committed:
department locks the artificial volite recipes

Wollywoger committed:
adds the techie wand sprite to the game

Wollywoger committed:
adds psionic quirk, fixes a runtime

Wollywoger committed:
changes psionic quirk's icon

Wollywoger committed:
adds light breaker

Wollywoger committed:
mild fixes, makes volite powder work

Wollywoger committed:
occulory changes, fixes volite powder for real

Wollywoger committed:
makes the altars craftable, some maintenance on the designs

Wollywoger committed:
typo

Wollywoger committed:
another typo yipee

Wollywoger committed:
mild oculory fixes & touchups

Wollywoger committed:
adds the meditation mat and all its functionality

Wollywoger committed:
Merge branch 'master' into magic-rework-part-2

Wollywoger committed:
fixes the typing on meditation mat, which makes it work as intended

Wollywoger committed:
Merge branch 'magic-rework-part-2' of https://www.github.com/wollywoger/maplestationcode into magic-rework-part-2

Wollywoger committed:
removes some temp code from mana tap

Wollywoger committed:
don't need that

Wollywoger committed:
adds meditation guidebooks and refactors the volitious lignite multipack

Wollywoger committed:
small changes, adds crafting for the mat and better hints at it

Wollywoger committed:
mild changes to mana charge spells, fixes a path error in meditation book

Wollywoger committed:
minor description/feedback changes from preround

Wollywoger committed:
kills an annoying run time

Wollywoger committed:
SOMEHOW FIXES LEYLINE TEETERING AND DECIMAL MESSES?????

Wollywoger committed:
adds sprites and better desc to the meditation mat

Wollywoger committed:
attempted to make early magic tech work, makes it default cause it didn't

Wollywoger committed:
gets early magic working as intended (thanks jade)

Wollywoger committed:
slight changes to lens and tech wand

Wollywoger committed:
fixes a number of issues from last night's session

Wollywoger committed:
fixes dme

Wollywoger committed:
minor linter fix

Wollywoger committed:
Merge branch 'magic-rework-part-2' of https://www.github.com/wollywoger/maplestationcode into magic-rework-part-2

Wollywoger committed:
is this what you want?

Wollywoger committed:
housekeeping surronding the meditate family, moves them to a subtype of spell/charged/

Wollywoger committed:
disables every pool taking from leylines automatically cause that has notable issues for the system (and some performance, ew)

Wollywoger committed:
completely re-does the ruleset system, drops cost of the ruleset system down to less than half of its previous

Wollywoger committed:
uhhh lets see if this'll fix it

Wollywoger committed:
should fix the linter error on the stump, adds anchor/unanchor to the stump as well

Wollywoger committed:
just... pretty much everything i've been working on for the past few weeks

Wollywoger committed:
adds magically blank, a quirk which makes you unable to cast magic normally, also adds a check during pool deletion

Wollywoger committed:
removes signal handler

Wollywoger committed:
quick postround fix for the cargo packs

Wollywoger committed:
does most of the stuff discovered after a "recent" playtest + other fix ups

Wollywoger committed:
mild fix, left hand assignment on a line

Wollywoger committed:
i'm stupid

Wollywoger committed:
fixes, repaths, and touch-ups

Wollywoger committed:
forgot to add and update the sprites for the oculory

Wollywoger committed:
Merge branch 'master' into magic-rework-part-2

Wollywoger committed:
adds a tooltip for using wands

Wollywoger committed:
FASTEST FIX IN THE WEST

Wollywoger committed:
Merge branch 'master' into magic-rework-part-2

MrMelbert commented on maplestation_modules/code/datums/quirks/neutral.dm:
```diff
@@ -287,3 +287,45 @@
+
+	if(carbon_quirk_holder.mana_pool) // if we already have a pool, get rid of it, which the player should probably have one at this moment
+		var/datum/mana_pool/pool_to_destroy = carbon_quirk_holder.mana_pool
+		QDEL_NULL(pool_to_destroy)
```
```suggestion
	QDEL_NULL(carbon_quirk_holder.mana_pool)
```
This is safe

MrMelbert commented on maplestation_modules/code/datums/quirks/neutral.dm:
```diff
@@ -287,3 +287,45 @@
+
+	if(carbon_quirk_holder.mana_pool) // this check is absolutely a failsafe because this quirk implies you'll have a pseudo-pool to begin with, but this check is just in case you some how got it deleted
+		var/datum/mana_pool/pool_to_destroy = carbon_quirk_holder.mana_pool
+		QDEL_NULL(pool_to_destroy)
```
```suggestion
	QDEL_NULL(carbon_quirk_holder.mana_pool) 
```
Same

MrMelbert commented on maplestation_modules/code/datums/quirks/neutral.dm:
```diff
@@ -287,3 +287,45 @@
+		QDEL_NULL(pool_to_destroy)
+
+	var/datum/mana_pool/mob/living/carbon/blank/pool_to_add = new()
+	carbon_quirk_holder.mana_pool = pool_to_add
```
```suggestion
	carbon_quirk_holder.mana_pool = new /datum/mana_pool/mob/living/carbon/blank()
```

MrMelbert commented on maplestation_modules/code/datums/quirks/neutral.dm:
```diff
@@ -287,3 +287,45 @@
+
+	var/datum/mana_pool/pool_to_add = new type()
+
+	carbon_quirk_holder.mana_pool = pool_to_add
```
```suggestion
	var/datum/mana_pool/pool_type = carbon_quirk_holder.get_initial_mana_pool_type()

	carbon_quirk_holder.mana_pool = new pool_type()
```
Just in case (to avoid conflict with `src.type`

MrMelbert commented on maplestation_modules/code/modules/magic/mana/sources/altars/_magic_altars.dm:
```diff
@@ -0,0 +1,108 @@
+		if (!(pool_owner.loc))
+			return FALSE
+		else
+			target_pool_owner = pool_owner
```
```suggestion
		if (isnull(pool_owner.loc))
			return FALSE
		target_pool_owner = pool_owner
```

MrMelbert commented on maplestation_modules/code/modules/magic/mana/sources/altars/_magic_altars.dm:
```diff
@@ -0,0 +1,108 @@
+/obj/structure/magic_altar/wrench_act(mob/living/user, obj/item/tool)
+	. = ..()
+	default_unfasten_wrench(user, tool, 5 SECONDS)
+	return ITEM_INTERACT_SUCCESS
```
```suggestion
	if(default_unfasten_wrench(user, tool, 5 SECONDS))
		return ITEM_INTERACT_SUCCESS
```

MrMelbert commented on maplestation_modules/code/modules/magic/mana/sources/altars/_magic_altars.dm:
```diff
@@ -0,0 +1,108 @@
+	return /datum/mana_pool/magic_altar
+
+/obj/machinery/power/magic_contraption/wrench_act(mob/living/user, obj/item/tool)
+	. = ..()
```
```suggestion
```

MrMelbert commented on maplestation_modules/code/modules/magic/mana/sources/altars/_magic_altars.dm:
```diff
@@ -0,0 +1,108 @@
+		return ITEM_INTERACT_SUCCESS
+
+/obj/machinery/power/magic_contraption/crowbar_act(mob/living/user, obj/item/tool)
+	. = NONE
```
```suggestion
```

MrMelbert commented on maplestation_modules/code/modules/magic/mana/sources/altars/meditation_mat.dm:
```diff
@@ -0,0 +1,108 @@
+/obj/machinery/meditation_mat/proc/apply_effects(mob/living/target)
+	if(target != occupant)
+		return
+	target.apply_status_effect(/datum/status_effect/meditation_mat, MEDITATION_MAT_EFFECT)
+
+/obj/machinery/meditation_mat/proc/remove_effects(mob/living/target)
+	target.remove_status_effect(/datum/status_effect/meditation_mat, MEDITATION_MAT_EFFECT)
```
`MEDITATION_MAT_EFFECT` is doing nothing here because you aren't using a grouped status effect and don't have an `on_creation` override to accept it

MrMelbert commented on maplestation_modules/code/modules/magic/mana/sources/altars/nature_shrine.dm:
```diff
@@ -0,0 +1,88 @@
+		return
+	if (is_type_in_typecache(sacrifice, nature_shrine_mana_low))
+		accept_sacrifice(sacrifice, nature_shrine_mana_low)
+		return
```
Parent call either does nothing here or should be respected by checking return value

Additionally, you probably want to `return ITEM_INTERACT_SUCCESS` on accepting sacrifice

MrMelbert commented on maplestation_modules/code/modules/magic/mana/sources/altars/nature_shrine.dm:
```diff
@@ -0,0 +1,88 @@
+		return
+
+/obj/structure/magic_altar/nature/proc/accept_sacrifice(obj/item/sacrifice, mana_value)
+	QDEL_NULL(sacrifice)
```
`qdel(sacrifice)` is all you need, though I will say just deleting the sacrifice is a bit tame (I would add a sound effect, a chat message, and/or maybe leave behind some ash)

MrMelbert commented on maplestation_modules/code/modules/magic/mana/sources/altars/nature_shrine.dm:
```diff
@@ -0,0 +1,88 @@
+/obj/structure/magic_altar/nature/crowbar_act(mob/living/user, obj/item/tool)
+	balloon_alert(user, "deconstructing stump...")
+	if(!tool.use_tool(src, user, 5 SECONDS, volume=50))
+		return
```
```suggestion
		return ITEM_INTERACT_BLOCKING
```

MrMelbert commented on maplestation_modules/code/modules/magic/mana/sources/altars/nature_shrine.dm:
```diff
@@ -0,0 +1,88 @@
+	tool.play_tool_sound(src)
+	new /obj/item/stack/sheet/mineral/wood(get_turf(src), drop_amount)
+	new /obj/item/mana_battery/mana_crystal/standard(get_turf(src), 1)
+	qdel(src)
```
```suggestion
	deconstruct(TRUE)
```

MrMelbert replied:
Bonus points to put code for dropping stuff in `on_deconstruct`

MrMelbert commented on maplestation_modules/code/modules/magic/mana/sources/altars/nature_shrine.dm:
```diff
@@ -0,0 +1,88 @@
+			return ITEM_INTERACT_SUCCESS
+		if(FAILED_UNFASTEN)
+			return ITEM_INTERACT_BLOCKING
+	return .
```
If I'm not mistaken you alreay have unfastening code for `/obj/structure/magic_altar` (the parent) so this is redundant / duplicate code here

MrMelbert commented on maplestation_modules/code/modules/magic/mana/sources/altars/stellar_oculory.dm:
```diff
@@ -0,0 +1,126 @@
+		active = TRUE
+		icon_state = "stellar"
+		return balloon_alert(user, "activated the oculory")
+
```
To my knowledge the return value for balloon alert is nothing so rather than `return balloon_alert()` we should be `return CLICK_ACTION_SUCCESS`

MrMelbert commented on maplestation_modules/code/modules/magic/mana/sources/altars/stellar_oculory.dm:
```diff
@@ -0,0 +1,126 @@
+	if(!active)
+		active = TRUE
+		icon_state = "stellar"
+		return balloon_alert(user, "activated the oculory")
```
This item state handling should be in `update_icon_state`

MrMelbert commented on maplestation_modules/code/modules/magic/mana/sources/altars/stellar_oculory.dm:
```diff
@@ -0,0 +1,126 @@
+		if(NO_STARLIGHT)
+			last_pulse_value = NO_STARLIGHT
+			pulse_value = low_pulse_value
+			icon_state = "stellar"
```
Same for this icon state handling (or use `flick` IG)

MrMelbert commented on maplestation_modules/code/modules/magic/mana/sources/altars/stellar_oculory.dm:
```diff
@@ -0,0 +1,126 @@
+/obj/machinery/power/magic_contraption/stellar_oculory/examine(mob/user)
+	. = ..()
+	if(!active)
+		return . += "The oculory can be turned on using Alt-Click."
```
```suggestion
		. += span_notice("[src] can be turned on using <b>Alt-Click</b>.")
		return
```

MrMelbert commented on maplestation_modules/code/modules/magic/mana/sources/altars/stellar_oculory.dm:
```diff
@@ -0,0 +1,126 @@
+			. += "The oculory is functioning under normal conditions."
+		if(NO_STARLIGHT)
+			. += span_warning("The oculory needs to gather more starlight.")
+	return . += "The oculory can be shut off using Alt-Click."
```
```suggestion
	. += span_notice("[src] can be shut off using <b>Alt-Click</b>.")
```

MrMelbert commented on maplestation_modules/code/modules/magic/mana/sources/altars/stellar_oculory.dm:
```diff
@@ -0,0 +1,126 @@
+		if(PARTIAL_STARLIGHT)
+			. += "The oculory is functioning under normal conditions."
+		if(NO_STARLIGHT)
+			. += span_warning("The oculory needs to gather more starlight.")
```
```suggestion
		if(FULL_STARLIGHT)
			. += span_warning("[src] is beaming with starlight!")
		if(PARTIAL_STARLIGHT)
			. += "[src] is functioning under normal conditions."
		if(NO_STARLIGHT)
			. += span_warning("[src] needs to gather more starlight.")
```

MrMelbert commented on maplestation_modules/code/modules/magic/mana/sources/altars/stellar_oculory.dm:
```diff
@@ -0,0 +1,126 @@
+		user.balloon_alert(user, "deactivate first!")
+		return .
+	if(default_deconstruction_screwdriver(user, "stellar_t", "stellar", tool))
+		return ITEM_INTERACT_SUCCESS
```
```suggestion
/obj/machinery/power/magic_contraption/stellar_oculory/screwdriver_act(mob/living/user, obj/item/tool)
	if (active)
		user.balloon_alert(user, "deactivate first!")
		return ITEM_INTERACT_BLOCKING
	if(default_deconstruction_screwdriver(user, "stellar_t", "stellar", tool))
		return ITEM_INTERACT_SUCCESS
```
Also should move this above the circuitborard definition

MrMelbert commented on maplestation_modules/code/modules/magic/mana/living_mana.dm:
```diff
@@ -55,3 +57,21 @@
+		else if (amount > softcap)
+			general_amount_estimate = "OVERLOADED"
+	else
+		general_amount_estimate = "ERROR"
+
+	status_tab += "Mana Count: [general_amount_estimate]"
```
Does this just permanently output `Mana count: ERROR`? Seems odd

MrMelbert commented on maplestation_modules/code/modules/magic/mana/magic_wands.dm:
```diff
@@ -0,0 +1,145 @@
+	var/datum/mana_pool/wand_pool = src.mana_pool
+
+	if(!target_mana_pool)
+		return  // no response for this failing, as else it would proc on ~70% of things in the codebase
```
Add return values to this proc, i.e. `return NONE` for no interaction / `return ITEM_INTERACT_BLOCKING` for failed interactions / `return ITEM_INTERACT_SUCCESS` for successful interactions

MrMelbert commented on maplestation_modules/code/modules/magic/story_spells/mana_charge.dm:
```suggestion
	cast_on.bleed(blood_drain)
```
`bleed` handles leaving a pool of blood for you want whatnot

MrMelbert commented on maplestation_modules/code/__HELPERS/object_helpers.dm:
```diff
@@ -0,0 +1,76 @@
+	if(starlight_level_from_above > current_starlight_level)
+		return starlight_level_from_above
+	else
+		return current_starlight_level
```
```suggestion
	return max(current_starlight_level, starlight_level_from_above)
```
Waow

MrMelbert commented on maplestation_modules/code/__HELPERS/object_helpers.dm:
```diff
@@ -0,0 +1,76 @@
+		if(PARTIAL_STARLIGHT)
+			return PARTIAL_STARLIGHT
+		if(FULL_STARLIGHT)
+			return FULL_STARLIGHT
```
```suggestion
	var/tile_starlight = checktilestarlight(turf_of_target, FALSE, starlight_max_range)
	if(tile_starlight)
		return tile_starlight
```

Wollywoger committed:
Apply the most straightforward suggestions from code review

did the stuff that is just straightforward changes; will address the stuff that requires more complex changes in a later batch

Co-authored-by: MrMelbert <51863163+MrMelbert@users.noreply.github.com>

Wollywoger committed:
second batch of changes following suggestions
